### ISSUE TITLE
Add support for standardized plugin metadata in file (rockspec, `plugin.lua`, `requirements.lua`, etc.)

### ISSUE THREAD

**Issue by mjlbach** [wbthomason/packer.nvim#808]
Plugins cannot currently declare their own dependencies. While this has historically been fine for vim plugins (at least) because plugins were mostly self-contained, many plugins now have dependencies on other plugins (plenary, nui, telescope, etc.). We should have a standard format for declaring these dependencies (with version constraints, if applicable) along with the minimum/maximum neovim version supported.

[Rockspec](https://github.com/stevedonovan/rockspec/blob/master/docs/readme.md) seems like a natural choice.

The reasoning I'm opening the issue here, is I think it is probably easier to get a critical mass of plugin maintainers (namely me, @lewis6991 and @tjdevries) to agree to something, add support to packer, and then convince others to join in, then other package managers can follow suit.

**Comment by wbthomason** [comment]
A few notes:
- There's some prior discussion of related functionality here: https://github.com/wbthomason/packer.nvim/discussions/176
- Rockspec has the advantage of standardization, but (I think) may be more complex than we need (e.g. including build steps, etc.). I don't want to reinvent the wheel, but I would lean slightly toward a simpler format that's either JSON (familiar to many developers, easy to parse in Neovim) or a Lua module (trivial to import as a table). I could be wrong, but I think we'd need a rockspec parser if we went that route (though I am not super familiar with the details of writing rockspecs).
- We should include fields for:
  - Plugin name
  - Plugin version
  - List of plugin dependencies, with version constraints (need to discuss what subset of version constraints we can support)
  - Maybe settings with defaults? Could do some interesting things with config autocompletion with this
  - What else?
- I think we can get a basic constraint solver implemented in `packer` without too much trouble. The last time I looked into using Luarocks as a library (in any part), the impression I got was that it was impossible, but this may have changed.

**Comment by mjlbach** [comment]
Json has the advantage that it's easier to parse from other language than lua, where you'd have to reimplement a lua table parser.

What would be the concrete steps that would be needed from the spec side to make this happen? 

> 
    Plugin name
    Plugin version
    List of plugin dependencies, with version constraints (need to discuss what subset of version constraints we can support)
    Maybe settings with defaults? Could do some interesting things with config autocompletion with this
    What else?

Just put these in a template json file in `nvim-lua` and say "this is the tentative plugin spec v0.0.1"?

**Comment by Shougo** [comment]
I think JSON is better.

**Comment by lewis6991** [comment]
For me JSON is a big no:
- It doesn't support comments. This on its own is enough reason not to use it for **any** source.
- It doesn't allow any form of dynamic behaviour which are going to want, e.g:
    - Conditional dependencies for platform/os/neovim version.
    - Debug/test versions of a plugin will need additional meta-data.

Yes, it has the benefit of being easy to parse, but Lua is just as easy. If non-Lua plugins want to use this (I can't see a reason why they would, only package managers do), then they can just have some small Lua code to parse and elaborate the file and get at it via RPC, we could even provide that via a standard API.

In regard to rockspec vs requirements.lua, I agree rockspec is more than we require right now, but that doesn't prevent using a subset of rockspec to begin with. It also doesn't stop us adding Neovim specific data. If we use something that is rockspec compatible in some form, we may eventually be able to leverage luarocks in the future. I would imagine we would want to stop using git repo's for plugin downloading at some point?

Other examples for us to learn from: [rockspec (lua)](https://github.com/luarocks/luarocks/wiki/Rockspec-format), [setup.py (python)](https://docs.python.org/3/distutils/setupscript.html), [build.sbt (scala)](https://www.scala-sbt.org/1.x/docs/Basic-Def.html), [go.mod (go)](https://go.dev/ref/mod). They all do the same thing.

**Comment by danymat** [comment]
I know that the WIP plugnplay repo (https://github.com/nvim-plugnplay/plugnplay.nvim) is working with json or toml configuration. I think that toml can be a good idea, and personally, better than json.

**Comment by max397574** [comment]
We already had the idea of asking plugin devs to do something like this for plugnplay.
The data we would like to have:
- name
- plugin version
- dependencies (nvim-lua/plenary.nvim (repo adress))
- code needed so setup plugin with the defaults
optional but would be really cool:
- categories (we need to define some categories like completion, lsp, file explorer which people could use)
this would make stuff like a plugin library (which we'd like to do) possible

It doesn't rly matter if it's in json or toml
We use jsonc for our plugin configurations because it's easy to parse and can have comments

**Comment by lewis6991** [comment]
Why not lua?

**Comment by max397574** [comment]
I don't know how hard it would be to parse lua from other languages
I'm not the one that had the idea to use json xD (@vhyrro)

**Comment by lewis6991** [comment]
I don't think being able to parse from other languages is really a requirement. Neovim will always have lua and so will always be able to parse lua and provide it in whatever form any RPC plugins wants.

**Comment by max397574** [comment]
I think lua could be useful for this because we could for example do this:
```lua
default_setup = function() require"plugin".setup() end
```
This could be used to provide ootb experience for plugins

**Comment by Shougo** [comment]
Of course, I like toml than JSON.
But the compatibility is very important for the metadata format
Toml parsing is not standard.
And the Lua format is for only Lua (neovim) plugins.

**Comment by lewis6991** [comment]
> And the Lua format is for only Lua (neovim) plugins.

As I mentioned, Neovim will be able to parse a lua format directly and provide the data via RPC to non-Lua plugins.

**Comment by mjlbach** [comment]
Let me just nip a few things in the bud.

* Neovim is unlikely to bundle a toml parser
* I am unlikely to add jsonc support to the json parser
* The advantage of json is that most languages (including javascript most obviously) can directly process it as opposed to lua, where they would most likely have to spawn a neovim instance to parse the lua.
* I don't think a design goal should be to programmatically change the configuration, I think (like rockspec) it should be a simple lua table. I would go so far as to say we set the local function environment to disable function evaluation.

> code needed so setup plugin with the defaults

I don't think this is a good idea, or at least, it's not the first step so I would rather we remained focus on creating a minimal viable first draft of the specification.

**Comment by Shougo** [comment]
The format should not be for packer.nvim only.
So I think JSON is better.
The package manager is not only packer.nvim.
The format file is included in plugin repository isn't it?
I don't want to include lua file in non Lua plugins.

**Comment by max397574** [comment]
> I don't think this is a good idea, or at least, it's not the first step so I would rather we remained focus on creating a minimal viable first draft of the specification.

ofc it isn't the first thing to do
but it would be really cool to have this *somewhen*

**Comment by kyoh86** [comment]
I believe that simple goals should have simple solutions.
lua is too rich for expressing metadata.

**Comment by kuuote** [comment]
In my opinion, shouldn't support dynamic behavior. It's not a role of meta data.

**Comment by shadmansaleh** [comment]
Ohh we are doing lua vs json again 😅.

In my opnion this format shouldn't be specific to packer or just neovim . Vim should be able to easily parse it too so vim-plug, dein ... can add support for same specification file easily (Instead of each plugin manager having it's own spec) .

**Comment by mjlbach** [comment]
While I also lean towards json, I also do not think explicit vim/vimscript only compatibility should be a goal, as vim/neovim will continue to diverge after vim 9. Json does have the benefit that both editors have built-in support.

**Comment by ii14** [comment]
What about JSON5, YAML, TOML?

**Comment by lewis6991** [comment]
Since the argument is between JSON and Lua, can we just support both and be done with it?

**Comment by mjlbach** [comment]
Yeah, I think it's trivial to support both. Packages can provide either/or and it is trivial to interconvert.

**Comment by wbthomason** [comment]
I agree - the "hard" part of supporting metadata is not parsing the format. Let's try to agree on a schema, and that can be supported as JSON/Lua/whatever future format.

**Comment by mjlbach** [comment]
I made a proposal on the repo I created here, I'll give @lewis6991 and @wbthomason write access

https://github.com/nvim-lua/nvim-package-specification/pull/1/files

**Comment by danymat** [comment]
> I made a proposal on the repo I created here, I'll give @lewis6991 and @wbthomason write access
>
> https://github.com/nvim-lua/nvim-package-specification/pull/1/files

(Disclaimer: I'm sorry if I sound rude, I'm not English fluent and I don't have a lot of vocabulary to pick words from 😞)

I don't think this is a good idea:
- Why do you want to externalize the discussion of providing plugin standardization for private views only ?
- And so, why including only a very small number of participants in it ? Are the participants you invited only because you like what they do and provide you great feedback ? 
- If so, what about the feedback other plugin contributors are giving you ?

If you want to make a proposal, I suggest:
- Make your proposal here so that anyone can view and participate
- Make your proposal in your repo, but let people have a way to participate: discussions, PR, Issues

**Comment by mjlbach** [comment]
> And so, why including only a very small number of participants in it ? Are the participants you invited because you like what they do and provide you great feedback, or because they have something other contributors don't have ?

Anyone can participate, the repo was private for ~5 minutes because it defaulted to private when I created it from the GH ui, posted about it, then changed the visibility settings from the GH interface. There wasn't anything nefarious.

> And so, why including only a very small number of participants in it ? Are the participants you invited because you like what they do and provide you great feedback, or because they have something other contributors don't have ?

Everyone is included, but I only gave *write* access to people I have explicitly worked with in the past.

> Make your proposal here so that anyone can view and participate

I did make my proposal, but the proposal itself is beyond packer and should not live in the packer repo. This issue is about implementing support for the standard, now there is a concrete PR that proposes the standard itself.

> Make your proposal in your repo, but let people have a way to participate: discussions, PR, Issues

See point 1.

**Comment by mjlbach** [comment]
There are a few remaining issues, but `packspec` is almost complete and nearing the first release of the specification. You can peruse/provide commentary here: https://github.com/nvim-lua/nvim-package-specification

The main unresolved items are:

* the dependency format, which may change
* the build step/post-install hooks
